### PR TITLE
Respawn self-play workers to avoid memory leak

### PR DESCRIPTION
**Pull request by jonbinney**
Adds a bit of overhead but does fix the leak! `train_alphazero.py` was getting quite large, so I moved the worker starting/restarting/etc to its own class in self_play_manager.py. In this PR I also updated the ruff and pylance settings so that I don't keep missing type warnings.

A few of other minor things:
- updated the check in AlphaZeroAgent.train_iteration to make sure that the replay buffer is longer in batch size even after pulling out the validation set
- fixed the model path passed to the benchmarking
- switched to calling AlphaZeroAgent.train_iteration instead of calling NNEvaluator.train_iteration directly from train.py, because that was bypassing the code that saves the replay buffer.

### CODE REVIEW

**Comment by jonbinney** [comment]
This PR isn't as big and scary as it looks - most of it is moving the self-play code from train_alphazero.py to self_play_manager.py.
